### DISCUSSION TITLE
Feedback after porting ~80 shaders from GLSL to WGSL

### DISCUSSION THREAD

**Discussion by AshleyScirra** [gpuweb/gpuweb#2204] [Ideas/Feedback]
I just finished porting the ~80 shader effects supported in Construct from GLSL to WGSL. It took several working days to complete. I wrote down some notes about the process I thought would be worth sharing with the people involved in standardisation. I worked in Chrome Canary and I might have noted things that are specified but not yet implemented in Chrome.

Much of this has probably been discussed elsewhere (or even things I brought up elsewhere), but I'm repeating it anyway as it's all relevant to the experience of porting GLSL to WGSL.

In short WGSL has a nice syntax, but is far more verbose than GLSL. The approach to things like variable types and functions is nice. But you often have to take far more roundabout ways of writing relatively simple code in WGSL.

### Repeating types
As noted the way variable types work is nice - even if it takes a lot of typing - but it feels egregious when having to repeat types in initialisers, e.g.:

```
var myVec : vec2<f32> = vec2<f32>(0.0, 1.0);
```

This declaration has to repeat the type `vec2<f32>` twice, and this kind of thing comes up all the time. It also makes it easy to make a mistake when changing a variable type; you have to remember to update both types, and I'd often find myself updating just one and forgetting the other. It would be great if we could omit the type, or write something like `auto` instead, and have it deduce the type from the right hand side, e.g.:

```
// option 1 - omit type, automatically deduce
var myVec = vec2<f32>(0.0, 1.0);

// option 2 - use auto type
var myVec : auto = vec2<f32>(0.0, 1.0);
```

### Ternary ?: operator
I miss the ternary operator in WGSL. Here's some GLSL code for an overlay blend using ?: for three components:

```
front.r = back.r < 0.5 ? 2.0 * back.r * front.r : 2.0 * (front.r + back.r * front.a - back.r * front.r) - front.a;
front.g = back.g < 0.5 ? 2.0 * back.g * front.g : 2.0 * (front.g + back.g * front.a - back.g * front.g) - front.a;
front.b = back.b < 0.5 ? 2.0 * back.b * front.b : 2.0 * (front.b + back.b * front.a - back.b * front.b) - front.a;
```

In WGSL this has to be expanded to:

```
if (back.r < 0.5)
{
	front.r = 2.0 * back.r * front.r;
}
else
{
	front.r = 2.0 * (front.r + back.r * front.a - back.r * front.r) - front.a;
}

if (back.g < 0.5)
{
	front.g = 2.0 * back.g * front.g;
}
else
{
	front.g = 2.0 * (front.g + back.g * front.a - back.g * front.g) - front.a;
}

if (back.b < 0.5)
{
	front.b = 2.0 * back.b * front.b;
}
else
{
	front.b = 2.0 * (front.b + back.b * front.a - back.b * front.b) - front.a;
}
```

This feels unnecessarily verbose and makes it harder to see at a glance that the same thing is being done on each channel.

### Mandatory braces
The mandatory braces around if/else also contribute to making code verbose. It would be nice if the above code could at least be written like this:

```
if (back.r < 0.5)
	front.r = 2.0 * back.r * front.r;
else
	front.r = 2.0 * (front.r + back.r * front.a - back.r * front.r) - front.a;

if (back.g < 0.5)
	front.g = 2.0 * back.g * front.g;
else
	front.g = 2.0 * (front.g + back.g * front.a - back.g * front.g) - front.a;

if (back.b < 0.5)
	front.b = 2.0 * back.b * front.b;
else
	front.b = 2.0 * (front.b + back.b * front.a - back.b * front.b) - front.a;
```

But WGSL rules currently prevent this.

### No l-value swizzling
You can't assign to a reduced set of vector components, even in order, like you can in WGSL. For example WGSL does not allow the following:

```
var color : vec4<f32> = vec4<f32>(0.0, 0.0, 0.0, 0.0);
// ...
color.rgb = otherColor.rgb;  // not allowed
```

Again WGSL requires a more verbose form again repeating the type:

```
color.rgb = vec4<f32>(otherColor.rgb, color.a);
```

### Missing arithmetic assignment operators
There are no `+=`, `*=` etc. operators in WGSL. This also requires more verbose code with lots of `x = x * 2` etc.

### Missing increment/decrement
There's no `++` and `--` in WGSL.
This makes for loops look a bit silly, as you have to write `i = i + 1` as `i++` is not allowed, nor even `i += 1`.

### Combining effects
Several of these things often combine to make elegant GLSL code awkward to port to WGSL. For example in GLSL to unpremultiply a color:

```
if (color.a != 0.0)
    color.rgb /= color.a;
```

In WGSL you have to write this:

```
if (color.a != 0.0)
{
    color = vec4<f32>(color.rgb / color.a, color.a);
}
```

### Inconsistent vector/scalar overloads
`vec3<f32> + f32` works, but comparison operators like `vec3<f32> < f32` don't work and has to instead be rewritten as `vec3<f32> < vec3<f32>`. This feels inconsistent, both within WGSL and with GLSL.

Similarly `pow(vec3<f32>, f32)` works in GLSL but not WGSL; `clamp(vec2<f32>, f32, f32)` works in GLSL but not WGSL; etc.

### Inconsistent mod/% operator
At first glance WGSL the `%` operator looks like the replacement for GLSL `mod()`, as WGSL does not appear to have an equivalent mod() function.

WGSL `x % y` does `x - y * trunc(x/y)`. GLSL `mod(x, y)` does `x - y * floor(x/y)`. `floor` is different to `trunc` for negative numbers. This caused a bit of a headache trying to port GLSL mod() to WGSL: `%` breaks if you use negative numbers, and it seems the easiest thing to do is just write your own `mod` function in WGSL.

### atan2 naming
GLSL lets you use `atan(x)` or `atan(y, x)`; WGSL calls the latter `atan2`. Just a minor thing that momentarily tripped me up.

### My WGSL wishlist
- ternary ?: operator
- 'auto' type
- optional braces
- arithmetic assignment operators e.g. *=
- increment and decrement operators ++, --
- better swizzling e.g. color.rgb = someVec3
- conversions for comparisons vec < scalar
- overloads for built-ins e.g. func(vec, scalar)
- keep compatibility with GLSL where possible: add a mod() function that works the same; use atan(y, x) instead of atan2; etc.

**Comment by ben-clayton** [comment]
Thank you for posting this. I think many of the issues you've mentioned already have github issues open., but we'll make sure that this is all tracked.

> ### Repeating types
> ```
> // option 1 - omit type, automatically deduce
> var myVec = vec2<f32>(0.0, 1.0);
> 
> // option 2 - use auto type
> var myVec : auto = vec2<f32>(0.0, 1.0);
> ```
 
Option 1 has been part of the spec and has been implemented in Chrome for some time now. Is this not working for you?

**Reply by AshleyScirra** [reply]
Wow, you're right, it does work! I simply had no idea it was supported. Wish I knew that before I started! I guess that's the downside of a lack of documentation at the moment. The spec is not a very good substitute for proper developer documentation (MDN style).

**Comment by kvark** [comment]
> Repeating types

As @ben-clayton mentioned, WGSL has type inference. You don't need to specify the type if there is an initializer.

> Ternary ?: operator

Use `select()` instead. It's almost equally short, and doesn't require special operator syntax and precedence rules.

> Mandatory braces

I think the group can agree that some of the braces are redundant, but last time I recall there wasn't a lot of appetite to change that.

> No l-value swizzling
> Missing arithmetic assignment operators

The group recognizes this as something available in all other languages. I think we consider it a minor annoyance, which is acceptable for V1. We can and probably should follow-up on this.

> Missing increment/decrement

I don't think this would be good to have, but open to other opinions. This requires new syntax, new parsing logic, new operator precedence rule, etc. Rust doesn't have ++ and --, fwiw.

> Combining effects

That's just a combination of "Missing arithmetic assignment operators" and "No l-value swizzling"

> Inconsistent vector/scalar overloads
> vec3<f32> + f32 works, but comparison operators like vec3<f32> < f32 don't work

I agree, I'm not a fan of this either. Would mildly prefer `vec3<f32> + f32` not being available (see #1520), but I imagine a lot of developers expect that to work.

> Inconsistent mod/% operator

It's actually consistent in WGSL in the way we agreed, which matches `OpFRem` and `%` in other languages. See #2187 for the missing spec piece.
Direct compatibility with GLSL is nice, but it's not a primary goal for WGSL design.

> atan2 naming

File a feature request? No strong preference from my side.

**Comment by munrocket** [comment]
Looks like a good tutorial, experienced the same. For me most confused was:
* explicit type conversion (vec3 + float) - because you need to visually check each expression.
* mod - because you need to test result (not just satisfy validator) or make helpers (mod1, mod2, mod3)

You not mentioned:
* #define -> global const, custom workflow
* smoothStep, faceForward, inverseSqrt

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2021-11-02-Minutes#feedback-after-porting-80-shaders-from-glsl-to-wgsl-2204">WGSL meeting minutes 2021-11-02</a></summary>



* MM: Really helpful list, thank you to the authors. Didn’t see anything that was super urgent, but together this is like a death by a thousand papercuts, so our feeling is that we want to fix *some* of these for v1. (presumably the easier ones)
* GR: Something that surprised me was the order of params in the `select` command. FWIW HLSL just got a `select` that is different: select(cond, true_val, false_val)
* …
* DN: MSL has (false, true, cond)
* JG: I would imagine that one reason to want cond last is because we want the instinct to be that authors expect true and false expressions to be both evaluated before choosing which one via cond.
* MM: In order for any of these issues to be fixed, we need filed Issues for them, so we should file Issues for them.
* GR: The `select` issue might actually only be from the blog post. There are slight differences between this Issue and the blog post
* 
* RM: Would rather have ‘if else’ as an expression.  Since they have braces, it would not have ambiguity.
* **GR to file `select` issue DONE: [Select operator arguments are in a non-intuitive order · Issue #2242 · gpuweb/gpuweb (github.com)](https://github.com/gpuweb/gpuweb/issues/2242)**
* **JG to file others**


</details>

**Comment by lf94** [comment]
Question: has "Inconsistent vector/scalar overloads" been corrected since this research?

After finding this and https://github.com/gpuweb/gpuweb/issues/610#issuecomment-699753812 , I want to use WGSL, and this inconsistency problem is one of the final things for me to decide on.

**Reply by lf94** [reply]
I suppose not https://gpuweb.github.io/gpuweb/wgsl/#arithmetic-expr

But reading the spec a bit, I think this may actually make more sense to keep it this way: `f32 < f32` or `vecN<f32> < vecN<f32>`, returning a scalar or vector boolean value respectively.
